contract_isp_invoice/models/contract.py

```diff
@@ -0,0 +1,535 @@
+        product_uom_obj = self.env['product.uom']
+        invoice_line_obj = self.env['account.invoice.line']
+        invoices = []
+        if self._context is None:
+            self._context = {}
+        if data is None:
+            data = {}
+
+        journal_types = {}
+
+        # prepare for iteration on journal and accounts
+        for line in self:
+            if line.journal_id.type not in journal_types:
+                journal_types[line.journal_id.type] = set()
+            journal_types[line.journal_id.type].add(line.account_id.id)
+        for journal_type, account_ids in journal_types.items():
+            for account in analytic_account_obj.browse(list(account_ids)):
+                # jgama - If there's a parent, invoice the parent
+                partner = account.partner_id.parent_id or account.partner_id
+
+                if (not partner) or not (account.pricelist_id):
+                    raise Warning(
+                        _('Analytic Account Incomplete!'),
+                        _('Contract incomplete. Please fill in the Customer'
+                          'and Pricelist fields.'))
+
+                date_due = False
+                if partner.property_payment_term:
+                    pterm_list = account_payment_term_obj.\
+                        compute(partner.property_payment_term.id, value=1,
+                                date_ref=time.strftime('%Y-%m-%d'))
+                    if pterm_list:
+                        pterm_list = [line[0] for line in pterm_list]
+                        pterm_list.sort()
+                        date_due = pterm_list[-1]
+
+                curr_invoice = {
+                    'name': time.strftime('%d/%m/%Y') + ' - ' + account.name,
+                    'origin': account.name,
+                    'partner_id': partner.id,
+                    'company_id': account.company_id.id,
+                    'payment_term': partner.property_payment_term.id or False,
+                    'account_id': partner.property_account_receivable.id,
+                    'currency_id': account.pricelist_id.currency_id.id,
+                    'date_due': date_due,
+                    'fiscal_position': partner.property_account_position.id
+                }
+                context2 = self._context.copy()
+                context2['lang'] = partner.lang
+                # set company_id in context, so the correct default journal
+                # will be selected
+                context2['force_company'] = curr_invoice['company_id']
+                # set force_company in context so the correct product
+                # properties are selected (eg. income account)
+                context2['company_id'] = curr_invoice['company_id']
+                last_invoice = invoice_obj.create(curr_invoice)
+                invoices.append(last_invoice.id)
+
+                self._cr.execute("""SELECT product_id, user_id, to_invoice,
```

Comment:
I am not comfortable with SQL statement.
There is currently no check on company_id which I think might be necessary